src/triples.rs

```diff
@@ -265,6 +269,123 @@ impl<'de> serde::Deserialize<'de> for TriplesBitmap {
+        for mut indices in indicess {
+            let mut first = true;
+            // sort by predicate
+            indices.sort_by_cached_key(|pos_y| wavelet_y.access(*pos_y as usize).unwrap());
+            for index in indices {
+                bitmap_index_bitvector.push_bit(first);
+                first = false;
+                cv.push_int(index as usize).unwrap();
+            }
+        }
+        let bitmap_index = Bitmap { dict: Rank9Sel::new(bitmap_index_bitvector) };
+        let op_index = OpIndex { sequence: cv, bitmap: bitmap_index };
+        debug!("built OPS index");
+        Self { order, bitmap_y, adjlist_z, op_index, wavelet_y }
+    }
+
+    /// Creates a new TriplesBitmap from a list of sorted RDF triples
+    // under construction, get it to compile with empty or nonsense values first, then fix
+    // originally written by Greg Hanson in the experimental-rdf2hdt PR, move here is in progress
+    pub fn from_triples(triples: Vec<TripleId>) -> Self {
+        let mut y_bitmap = BitVector::new();
+        let mut z_bitmap = BitVector::new();
+        let mut array_y = Vec::new();
+        let mut array_z = Vec::new();
+
+        let mut last_x = 0;
+        let mut last_y = 0;
+        let mut last_z = 0;
+
+        for (i, triple) in triples.iter().enumerate() {
+            let x = triple.subject_id;
+            let y = triple.predicate_id;
+            let z = triple.object_id;
+
+            if x == 0 || y == 0 || z == 0 {
+                panic!("triple IDs should never be zero")
+            }
+
+            if i == 0 {
+                array_y.push(y);
+                array_z.push(z);
+            } else if x != last_x {
+                if x != last_x + 1 {
+                    panic!("the subjects must be correlative.")
+                }
+
+                //x unchanged
+                y_bitmap.push_bit(true);
+                array_y.push(y);
+
+                z_bitmap.push_bit(true);
+                array_z.push(z);
+            } else if y != last_y {
+                if y < last_y {
+                    panic!("the predicates must be in increasing order.")
+                }
+
+                // y unchanged
+                y_bitmap.push_bit(false);
+                array_y.push(y);
+
+                z_bitmap.push_bit(true);
+                array_z.push(z);
+            } else {
+                if z < last_z {
+                    panic!("the objects must be in increasing order")
+                }
+
+                // z changed
+                z_bitmap.push_bit(false);
+                array_z.push(z);
+            }
+
+            last_x = x;
+            last_y = y;
+            last_z = z;
+        }
+
+        let bits_per_entry: usize = (triples.len().ilog2() + 1).try_into().unwrap();
+
+        let bitmap_y = Bitmap::new(y_bitmap.words().to_vec());
+        let bitmap_z = Bitmap::new(z_bitmap.words().to_vec());
+        let sequence_y = Sequence::new(&array_y, bits_per_entry);
```

Comment:
so far at least there is something wrong with the `sequence_y` value during the `convert` test in hdt.rs as the wavelet thread join is failing - haven't had a chance to investigate yet

Reply:
I will take a look.
I'm also not sure how much time the parallelization actually saves, as the criterion benchmark didn't show any measureable difference between joining the wavelet thread before or after loading the Z layer, but then the test dataset is not that large, but that is another topic.